force-app/main/default/classes/DMLOperations.cls

```diff
@@ -162,59 +196,103 @@ public with sharing class DMLOperations {
+            upsertAccount = newAccount;
+        }
 		// otherwise, create a new Account and set the description
-
+          
 		// Upsert the Account
-		return null;
-	}
-
-    /**
-     * Question 10
-	 * Manage the association between Contacts and Accounts
-	 * Ensure that for each Contact in the provided list, there is a corresponding Account Name that matches the Contact's last name.
-	 * If such an Account does not exist, the method creates it; if it does, the method ensures the Contact is associated with it. (Hint: use previous upsertAccount method - this may cause bad practices but for learning purposes it's okay.)
-     * Example Contact.LastName = 'Doe' Link to Account.Name = 'Doe' 
-     * Example Contact.LastName = 'Jane' Link to Account.Name = 'Jane'
-	 * @param contactsToUpsert List of Contacts to be upserted and linked to accounts.
-	 */
-	public static void upsertAccountsWithContacts(List<Contact> contactsToUpsert) {
-		// Loop through each Contact
-
-		// Extract the Account Name from Contact's LastName
-
-		// Get the account based on the name or create one if it doesn't exist
-
-		// Link the Contact to its Account using the Account Id
-
-		// Upsert the Contacts
+	      return upsertAccount;
 	}
+        /**
+         * Question 10
+         * Manage the association between Contacts and Accounts
+         * Ensure that for each Contact in the provided list, there is a corresponding Account Name that matches the Contact's last name.
+         * If such an Account does not exist, the method creates it; if it does, the method ensures the Contact is associated with it. (Hint: use previous upsertAccount method - this may cause bad practices but for learning purposes it's okay.)
+         * Example Contact.LastName = 'Doe' Link to Account.Name = 'Doe' 
+         * Example Contact.LastName = 'Jane' Link to Account.Name = 'Jane'
+         * @param contactsToUpsert List of Contacts to be upserted and linked to accounts.
+         */
+        public static void upsertAccountsWithContacts(List<Contact> contactsToUpsert) {
+            List<Account> accToStore = new List<Account>();
+            // Loop through each Contact
+            for(Contact contact : contactsToUpsert) {
+                String lastName = contact.LastName; // Corrected to LastName (not lastName)
+                Account upsertAccount;
+                //Query account with name that equals the last name
+                List<Account> matchingAccounts = [SELECT Id FROM Account WHERE Name = :lastName];
+                if(!matchingAccounts.isEmpty()) { // check if the last name equals the same as the account name
+                    upsertAccount = matchingAccounts[0];
+                } else {
+                    upsertAccount = new Account(Name = lastName); // This will create the Account with the same last name as the contact name
+                    accToStore.add(upsertAccount);
+                }
+            }
```

Comment:
You can't have SOQL within a for loop. What happens if we have 200 Contacts here? Use a single query and a map to find a better way to query/match your records.